DolbyIO/Source/Private/DolbyIOSubsystem.cpp

```diff
@@ -0,0 +1,637 @@
+// Copyright 2023 Dolby Laboratories
+
+#include "DolbyIOSubsystem.h"
+
+#include "DolbyIOLogging.h"
+#include "DolbyIOVideoSink.h"
+
+#if PLATFORM_MAC
+#define DOLBYIO_COMMS_SUPPRESS_APPLE_NO_RTTI_WARNING
+#endif
+#include <dolbyio/comms/sdk.h>
+
+#include "Async/Async.h"
+#include "Engine/Texture2D.h"
+#include "Engine/World.h"
+#include "GameFramework/PlayerController.h"
+#include "Misc/Base64.h"
+#include "TimerManager.h"
+
+void UDolbyIOSubsystem::Initialize(FSubsystemCollectionBase& Collection)
+{
+	Super::Initialize(Collection);
+
+#if PLATFORM_WINDOWS
+	dolbyio::comms::sdk::set_app_allocator(
+	    {::operator new,
+	     [](std::size_t Count, std::size_t Al) { return ::operator new(Count, static_cast<std::align_val_t>(Al)); },
+	     ::operator delete,
+	     [](void* Ptr, std::size_t Al) { ::operator delete(Ptr, static_cast<std::align_val_t>(Al)); }});
+#endif
+
+	ConferenceStatus = dolbyio::comms::conference_status::destroyed;
+
+	VideoSink = MakeShared<DolbyIO::FVideoSink>();
+
+	FTimerManager& TimerManager = GetGameInstance()->GetTimerManager();
+	TimerManager.SetTimer(LocationTimerHandle, this, &UDolbyIOSubsystem::SetLocationUsingFirstPlayer, 0.1, true);
+	TimerManager.SetTimer(RotationTimerHandle, this, &UDolbyIOSubsystem::SetRotationUsingFirstPlayer, 0.01, true);
+
+	OnTokenNeeded.Broadcast();
+}
+
+namespace
+{
+	std::string ToStdString(const FString& String)
+	{
+		return TCHAR_TO_UTF8(*String);
+	}
+
+	constexpr int ScaleCenti = 100;
+
+	FDolbyIOParticipantInfo ToUnrealParticipantInfo(const dolbyio::comms::participant_info& Info)
+	{
+		FDolbyIOParticipantInfo Ret{};
+		Ret.UserID = Info.user_id.c_str();
+		Ret.Name = Info.info.name.value_or("").c_str();
+		Ret.ExternalID = Info.info.external_id.value_or("").c_str();
+		Ret.AvatarURL = Info.info.avatar_url.value_or("").c_str();
+		Ret.bIsListener = Info.type && *Info.type == dolbyio::comms::participant_type::listener;
+		Ret.bIsSendingAudio = Info.is_sending_audio.value_or(false);
+		Ret.bIsAudibleLocally = Info.audible_locally.value_or(false);
+
+		FString DecodedExternalID;
+		FBase64::Decode(Ret.ExternalID, DecodedExternalID);
+		if (DecodedExternalID.StartsWith("{\"init-pos\": {\"x\": "))
+		{
+			Ret.bIsInjectedBot = true;
+
+			FString Left;
```

Comment:
I'm using Left and Right because that is what they name the parameters in https://docs.unrealengine.com/4.27/en-US/API/Runtime/Core/Containers/FString/Split .